internal/terraform/workspace/workspace_manager.go

```diff
@@ -0,0 +1,161 @@
+
+type WorkspaceFactory func(context.Context, string) (*workspace, error)
+
+type WorkspaceManagerFactory func(context.Context) WorkspaceManager
+
+func NewWorkspaceManager(ctx context.Context) WorkspaceManager {
+	return newWorkspaceManager(ctx)
+}
+
+func newWorkspaceManager(ctx context.Context) *workspaceManager {
+	wm := &workspaceManager{
+		ctx:    ctx,
+		ws:     make(map[string]*workspace, 0),
+		logger: defaultLogger,
+	}
+	wm.newWorkspace = wm.defaultWorkspaceFactory
+	return wm
+}
+
+func (wc *workspaceManager) defaultWorkspaceFactory(ctx context.Context, dir string) (*workspace, error) {
+	w := newWorkspace(ctx)
+
+	w.SetLogger(wc.logger)
+
+	d := &discovery.Discovery{}
+	w.tfDiscoFunc = d.LookPath
+	w.tfNewExecutor = exec.NewExecutor
+	w.newSchemaStorage = schema.NewStorage
+
+	w.tfExecPath = wc.tfExecPath
+	w.tfExecTimeout = wc.tfExecTimeout
+	w.tfExecLogPath = wc.tfExecLogPath
+
+	return w, w.init(ctx, dir)
+}
+
+func (wc *workspaceManager) SetTerraformExecPath(path string) {
+	wc.tfExecPath = path
+}
+
+func (wc *workspaceManager) SetTerraformExecLogPath(logPath string) {
+	wc.tfExecLogPath = logPath
+}
+
+func (wc *workspaceManager) SetTerraformExecTimeout(timeout time.Duration) {
+	wc.tfExecTimeout = timeout
+}
+
+func (wc *workspaceManager) SetLogger(logger *log.Logger) {
+	wc.logger = logger
+}
+
+func (wc *workspaceManager) AddWorkspace(dir string) error {
+	_, exists := wc.ws[dir]
+	if exists {
+		return fmt.Errorf("workspace %s was already added", dir)
+	}
+
+	w, err := wc.newWorkspace(wc.ctx, dir)
+	if err != nil {
+		return err
+	}
+
+	wc.ws[dir] = w
+
+	return nil
+}
+
+func (wc *workspaceManager) WorkspaceByPath(path string) (Workspace, error) {
+	if w, ok := wc.ws[path]; ok {
+		wc.logger.Printf("direct workspace lookup succeeded: %s", path)
+		return w, nil
+	}
+
+	dir := workspaceDirFromFilePath(path)
+	if w, ok := wc.ws[dir]; ok {
+		wc.logger.Printf("dir-based workspace lookup succeeded: %s", dir)
+		return w, nil
+	}
+
+	for _, w := range wc.ws {
+		if w.ReferencesModulePath(dir) {
+			wc.logger.Printf("module-ref-based workspace lookup succeeded: %s", dir)
+			return w, nil
+		}
+	}
+
+	return nil, &WorkspaceNotFoundErr{path}
+}
+
+func (wc *workspaceManager) ParserByPath(path string) (lang.Parser, error) {
+	w, err := wc.WorkspaceByPath(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return w.Parser(), nil
+}
+
+func (wc *workspaceManager) TerraformExecutorByPath(path string) (*exec.Executor, error) {
+	w, err := wc.WorkspaceByPath(path)
+	if err != nil {
+		return nil, err
+	}
+
+	return w.TerraformExecutor(), nil
+}
+
+// workspaceDirFromPath strips known lock file paths and filenames
+// to get the directory path of the relevant workspace
+func workspaceDirFromFilePath(filePath string) string {
+	for _, p := range pluginLockFilePaths("/") {
+		if strings.HasSuffix(filePath, p) {
+			return strings.TrimSuffix(filePath, p)
+		}
+	}
+
+	mlfp := moduleLockFilePath("/")
+	if strings.HasSuffix(filePath, mlfp) {
```

Comment:
Same as above, may need `filepath.Clean`

Reply:
As mentioned above - probably best to clean `filePath` once before even entering this function, instead of "re-cleaning" it downstream. The `moduleLockFilePath` also shouldn't require cleaning for the same reason as mentioned above (it's entirely static path just with OS-specific separators).